app/models/reducer.rb

```diff
@@ -43,60 +45,120 @@ def self.of_type(type)
+    if relevant_reductions.present?
+      extracts.map do |ex|
+        ex.relevant_reduction = if reduce_by_subject?
+            # find will only return the first match
+            relevant_reductions.find { |rr| rr.user_id == ex.user_id }
+          elsif reduce_by_user?
+            # find will only return the first match
+            relevant_reductions.find { |rr| rr.subject_id == ex.subject_id }
+          else
+            raise NotImplementedError.new 'This reduction topic is not supported'
+          end
       end
+    end
+
+    extracts
+  end
+
+  # load all reductions that might be relevant to any of the extracts that are
+  # being reduced
+  def get_relevant_reductions(extracts)
+    return [] unless user_reducer_keys.present? || subject_reducer_keys.present?
+
+    if reduce_by_subject?
+      UserReduction.where(user_id: extracts.map(&:user_id), reducible: reducible, reducer_key: user_reducer_keys)
+    elsif reduce_by_user?
+      SubjectReduction.where(subject_id: extracts.map(&:subject_id), reducible: reducible, reducer_key: subject_reducer_keys)
     else
-      reduction_fetcher.retrieve(reducer_key: key, subgroup: group_key).tap do |r|
-        # send empty data and store unless we're in running reduction mode
-        r.data = {}
-        r.store = {}
-      end
+      raise NotImplementedError.new 'This reduction mode is not supported'
     end
   end
 
+  # apply extract filters and deduplicate extracts (do not allow a running reduction
+  # to contain the same extract twice)
   def filter_extracts(extracts, reduction)
-    extracts = extracts.reject{ |extract| reduction.extract_ids.include? extract.id }
-    extract_filter.apply(extracts)
+    return extracts if extracts.blank?
+    extracts = extract_filter.apply(extracts)
+    extracts = extracts.reject do |extract|
+      reduction.extract_ids.include? extract.id
+    end if running_reduction?
+
+    extracts
+  end
+
+  # given an array of reductions to be updated, find or create one with the
+  # appropriate subgroup
+  def get_group_reduction(reductions, group_key)
+    match = reductions.find{ |reduction| reduction.subgroup == group_key }
+    if match.present?
+      match
+    else
+      if reduce_by_subject?
+        SubjectReduction.new \
+          reducible: reducible,
+          reducer_key: key,
+          subgroup: group_key,
+          subject_id: subject_id,
+          data: {},
+          store: {}
+      elsif reduce_by_user?
+        UserReduction.new \
+          reducible: reducible,
+          reducer_key: key,
+          subgroup: group_key,
+          user_id: user_id,
+          data: {},
+          store: {}
+      else
+        raise NotImplementedError.new 'This topic is not supported'
+      end
+    end
   end
 
   def associate_extracts(reduction, extracts)
+    # note that because we use deferred associations, this won't actually hit the database
+    # until the reduction is saved, meaning it happens inside the transaction
```

Comment:
The way this works is that we build up a set of reduction records, some of which exist in the database and are being updated, and some of which are novel, and then build associations for them to track which extracts have been used to produce those reductions. The deferred association gem means that the association records aren't saved until the transaction where the reductions are updated/created. otherwise rails would attempt to create those records immediately, which would fail in the case of novel reductions that aren't present in the database (because of the foreign key relationship)